### PR TITLE
[DEV-4081] Freshly baked signed cookies!

### PR DESCRIPTION
**Pull request by Elknar**
_This addresses:_
see internal PR

_By doing:_
New cookies are based on a counter
The counter is only reset when the session-store is reset, thus it is impossible to have session collision even in theory.
Security of the cookie is reliant on the SHA512-based HMAC
Cookie values are further obscured from the user with AES256 in CBC mode
Keys are randomly generated every time

_Major changes:_
Cookies now expire after 72hrs **NO MATTER WHAT**.
If someone manages to keep a session active the entire time then one of the following is true:

-  A poor human is being overworked to death - we should terminate the session out of mercy
-  Something failed with activity detection - we should terminate the session out of safety
-  People are hot-seating a computer or botting - we should terminate the session out of spite

The cookies are now also significantly longer - although this shouldn't pose an issue, as it is well within the standard limits

### CODE REVIEW

File: project.clj [hunk 1]
```diff
@@ -9,7 +9,14 @@
                               :username :env
                               :password :env}]]
 
-  :dependencies [[ring/ring-core "1.7.1"]]
+  :dependencies [[ring/ring-core "1.7.1"]
+                 [buddy/buddy-auth "2.2.0"]
```
Comment (MatthewDarling) [inline project.clj:13]: Do these various `buddy` dependencies mean we're now compatible with new or interesting tools? Like, are we "`buddy` compatible" such that somebody could more easily plug this library into their application?

I've never actually worked with `buddy`, so I guess I'm just wondering how our usage might interact with other people.

File: project.clj [hunk 2]
```diff
@@ -1,4 +1,4 @@
-(defproject kirasystems/aging-session "0.3.6-SNAPSHOT"
+(defproject kirasystems/aging-session "0.5.0-SNAPSHOT"
```
Comment (MatthewDarling) [inline project.clj:1]: What happened to `4.0`? :trollface:

File: src/aging_session/memory.clj [hunk 3]
```diff
@@ -1,10 +1,31 @@
 (ns aging-session.memory
   "In-memory session storage with mortality."
-  (:require [ring.middleware.session.store :refer :all])
-  (:import java.util.UUID))
+  (:require [ring.middleware.session.store :refer :all]
+            [ring.util.codec :refer [url-encode url-decode]]
+            [buddy.core.crypto :as crypto]
+            [buddy.core.mac :as mac]
+            [buddy.core.nonce :as nonce]
+            [clj-time.core :as time]
```
Comment (MatthewDarling) [inline src/aging_session/memory.clj]: `clj-time` is a transitive dep from `ring-core`. If they decide to switch to [`clojure.java-time`](https://github.com/dm3/clojure.java-time) (which everybody should be doing eventually), we'd have problems.

It might happen that somebody includes `aging-session` + `ring` in their project, then increases their Ring version, then suddenly our code produces a "class not found".

Anyway, the fix is easy: add `clj-time` as an explicit dependency. The `java-time` swap is in the works, so we can tackle it here when the time comes.

File: src/aging_session/memory.clj [hunk 4]
```diff
@@ -1,10 +1,31 @@
...
+            [clojure.tools.logging :as log]
+            [taoensso.nippy :as nippy])
+  (:import  [org.apache.commons.codec.binary Base64]))
 
 (defrecord SessionEntry [timestamp value])
 
+;; Base64 helpers
+(defn- string->bytes
+  "Convert UTF-8 string to bytes."
+  ^bytes [^String s]
```
Comment (MatthewDarling) [inline src/aging_session/memory.clj]: Huh, I thought return type hints [belonged before the function name](http://clojure-goes-fast.com/blog/performance-nemesis-reflection/). But the [official docs](https://clojure.org/reference/java_interop) show it this way:

> For function return values, the type hint can be placed before the arguments vector:
> 
>     (defn hinted-single ^String [])

I wonder if both versions work, or if the Clojure Goes Fast example is wrong... Ah well. Just wanted to confirm for anybody else curious that this is the right way.

File: src/aging_session/memory.clj [hunk 5]
```diff
@@ -43,6 +64,67 @@
...
 
+(def ^:private internals-atom (atom {:enc-key nil
+                                     :mac-key nil
+                                     :counter 0}))
+
+(defn get-crypto-keys
+  []
+  (let [{enc-key :enc-key
+         mac-key :mac-key} @internals-atom]
+    (if (and enc-key mac-key)                               ; if keys are present - use them
```
Comment (MatthewDarling) [inline src/aging_session/memory.clj]: I take it "keys" here means the `enc-key` and `mac-key` values. As opposed to the keys of the map structure. To disambiguate a bit, how about this wording:

```suggestion
    (if (and enc-key mac-key)                               ; if keys are initialized - use them
```

Admittedly, this mostly helps by making the distinction irrelevant - "does this refer to initializing real values instead of `nil`, or the encryption keys?" = yes :)

File: src/aging_session/memory.clj [hunk 6]
```diff
@@ -43,6 +64,67 @@
...
+      [enc-key mac-key]
+      (do                                                   ; otherwise generate new keys and attempt to set them
+        (swap! internals-atom update-in [:enc-key] #(if-not % (nonce/random-bytes 32) %))
```
Comment (MatthewDarling) [inline src/aging_session/memory.clj]: This is probably super edge case, but I think it would be good to make both changes in a single `swap!`. That way the atom state is either "both keys are `nil`" or "both keys are valid".

I thought it would be clever to do this using `merge-with`, but the argument order doesn't match well:

```clojure
(defn atomic-merge-with
  "For use with swap!, which always puts the atom value as the first
  argument, breaking merge-with."
  [atom-val fn-to-apply & args]
  (apply merge-with fn-to-apply atom-val args))

(swap! internals-atom atomic-merge-with
                      #(or %1 %2)
                      {:enc-key (nonce/random-bytes 32) :mac-key (nonce/random-bytes 128)})
```

So the best solution is probably to separate the computation from the stateful parts:

```clojure
(defn initialize-keys
  [internal-state]
  (if (and (:enc-key internal-state) (:mac-key internal-state))
    internal-state
    (merge-with #(or %1 %2)
                internal-state
                {:enc-key (nonce/random-bytes 32) :mac-key (nonce/random-bytes 128)})))

;;; inside get-crypto-keys
(swap! internals-atom initialize-keys)
```

Also, we can simplify the `merge-with` to plain `merge` if we don't need to handle the case where only one key is `nil`.

File: src/aging_session/memory.clj [hunk 7]
```diff
@@ -43,6 +64,67 @@
...
+      (log/error "Error decrypting cookie: " e)
+      nil)))
+
+(defn encrypt
+  [key data]
+  (let [iv (nonce/random-bytes 16)]
+    [iv (crypto/encrypt-cbc (crypto/block-cipher :aes :cbc) (nippy/freeze data) key iv)]))
+
+(defn verify-and-decrypt-cookie
+  [url-cookie]
```
Comment (MatthewDarling) [inline src/aging_session/memory.clj]: This function could use a docstring. Something to explain the theory behind the sequence of operations. My current understanding (without knowing too much of the crypto details):

  1. Decode the current cookie
  2. Retrieve the crypto keys
  3. Separate the cookie into component parts - `iv` is the salt, `enc-cookie` is the real cookie encrypted, `hmac` is (according to my quick search) a message authentication code
  4. Verify the encrypted cookie with the `hmac` value, binding the original value to the same name if verification succeeded, otherwise `nil`
  5. Decrypt the cookie if verification succeeded
  6. Pull out the cookie creation time as `iat`
  7. Return the cookie as valid if it's within 3 days of `iat`

Now that I've written out the steps and done a bit of research, I think I get it :) I'm a bit mystified by `iat`, but the other acronyms turned up fairly useful explanations in Google so they seem pretty standard.

Comment (MatthewDarling) [reply src/aging_session/memory.clj]: Anyway, I feel like a higher level summary is possible than just listing the steps like I did. That will help people to parse the code in the future.

File: src/aging_session/memory.clj [hunk 8]
```diff
@@ -43,6 +64,67 @@
...
+  (let [cookie (some-> url-cookie
+                       url-decode)
+        [enc-key mac-key] (get-crypto-keys)
+        [iv enc-cookie hmac] (string/split (or cookie "") #"\." 3)
+        enc-cookie (and hmac
```
Comment (MatthewDarling) [inline src/aging_session/memory.clj]: I found this `enc-cookie` binding kind of confusing on my first read. Since you're already taking advantage of the short circuiting of `and`, what do you think about eliminating this binding?

```clojure
(let [... earlier code ...
      cookie (and hmac (mac/verify ... ) (decrypt ... ))
      ... ]
  ... )
```

Done this way, it expresses verification as a natural step in the process of decrypting.

Comment (MatthewDarling) [reply src/aging_session/memory.clj]: Also, philosophical obersvation - I feel like there's a looooot of checks for `nil` in this function. `hmac = nil` leads to `enc-cookie  = nil` leads to `cookie = nil` leads to `iat = nil` leads to failing the `when` condition.

I can't think of a simple way to short circuit the larger `let` though. Having 3-4 functions with `when` on their inputs could be an alternative, but... I'm not sure that's better.

File: src/aging_session/memory.clj [hunk 9]
```diff
@@ -43,6 +64,67 @@
...
+        iat (and cookie
+                 (:iat cookie))]
+    (when (and cookie iat
+               (time/within? (time/interval (time/minus (time/now)
+                                                        (time/days 3))
+                                            (time/now))
+                             iat))
+      cookie)))
+
+(defn encrypt-and-hmac-cookie
```
Comment (MatthewDarling) [inline src/aging_session/memory.clj]: It might make sense to swap the order of the `encrypt` and `decrypt` function pairs in this namespace. Reading this function first explains `iat`, which is more of a mystery with the current ordering. Plus, you've got nothing to decrypt before you've got code to encrypt it.

File: src/aging_session/memory.clj [hunk 10]
```diff
@@ -53,22 +135,27 @@
...
+    (let [key (:id (verify-and-decrypt-cookie cookie))]
+      (swap! session-map sweep-entry event-fns key)
+      (when (and refresh-on-read (contains? @session-map key))
+        (swap! session-map assoc-in [key :timestamp] (now)))
+      (get-in @session-map [key :value] {})))
+
+  (write-session [_ cookie {:keys [bump-session?] :as data}]
+    (let [decrypted-cookie (verify-and-decrypt-cookie cookie)
+          key (or (:id decrypted-cookie)
+                  (:counter (swap! internals-atom update-in [:counter] inc)))
```
Comment (MatthewDarling) [inline src/aging_session/memory.clj]: Just so I understand better, when would we hit this branch?

  1. If `verify-and-decrypt-cookie` returns `nil`, we would hit this. That should happen if the cookie is edited. Are there other cases when it can happen, like having an empty cookie value? I see below that `encrypt-and-hmac-cookie` gets called if there's no valid decrypted cookie, so does this serve to grab the counter value when creating a new cookie?
  2. If `decrypted-cookie` is truthy, would it always have a truthy `:id` value?

File: src/aging_session/memory.clj [hunk 11]
```diff
@@ -53,22 +135,27 @@
...
+  (read-session [_ cookie]
+    (let [key (:id (verify-and-decrypt-cookie cookie))]
+      (swap! session-map sweep-entry event-fns key)
+      (when (and refresh-on-read (contains? @session-map key))
+        (swap! session-map assoc-in [key :timestamp] (now)))
+      (get-in @session-map [key :value] {})))
+
+  (write-session [_ cookie {:keys [bump-session?] :as data}]
+    (let [decrypted-cookie (verify-and-decrypt-cookie cookie)
+          key (or (:id decrypted-cookie)
```
Comment (MatthewDarling) [inline src/aging_session/memory.clj]: `key` seems like the wrong name for this, considering we have actual `enc-key` and `mac-key` and various other meanings in nearby code. `session-id`? If that's an overloaded term, `unique-id`?

File: test/aging_session/memory_test.clj [hunk 12]
```diff
@@ -5,51 +5,77 @@
...
+        (is (= (read-session as cookie) {:a 1}))
+        (is (not (= ts1 (read-timestamp as id))))
+        (is (= (read-session as cookie) {:a 1}))))))
+
+(deftest crypto-basic-functionality
+  (testing "can create keys"
+    (is (= 2 (count (get-crypto-keys)))))
+  (testing "key creation/retrieval is atomic"             ; this isn't proof, just a best-effort check
+    (is (= 1 (count (into #{} (doall (pmap #(do %& (map base64-encode (get-crypto-keys)))
+                                           (range 100))))))))
```
Comment (MatthewDarling) [inline test/aging_session/memory_test.clj]: Given the comment and this test case, I definitely think we should have a single `swap!` call. That way you don't write `enc-key` with one thread and `mac-key` with a different thread. I mean, their values are unrelated, so the outcome isn't really any different. But it's good to condense transactions.

**Review by MatthewDarling — DISMISSED** [review]
Feels like I left a lot of comments, but honestly a bunch of it is just questions so I understand better. In general, the flow makes sense - newly created cookies have the new format, we read existing cookies with the new format, we have a nice function which provides crypto keys and initializes them as needed... Makes sense to me 👍 

Summarizing the requested changes, in order of importance:

  1. Change `get-crypto-keys` to use a single `swap!`
  2. Add an explicit dependency on `clj-time`
  3. Verify the proper version number
  4. Add a docstring explaining `verify-and-decrypt-cookie`
  5. Tweak the first comment in `get-crypto-keys`
  6. Rename `key` inside `write-session` now that encryption is in play

Other than those + the questions, there's a couple opinionated style comments (function ordering + the `enc-cookie` shadowing), but those are optional.

File: project.clj (hunk 2 above)
Comment (Elknar) [reply project.clj:1]: Breaking change - originally was suggested to bump major version, but we'd go 0.x->1.x... so instead Joey suggested 0.5

File: src/aging_session/memory.clj (hunk 6 above)
Comment (Elknar) [reply src/aging_session/memory.clj]: I'm _really_ hesitant to do this. 

1. As you've stated lower, the two keys are unrelated - there's no difference in which thread sets them, we only care that they are set once. There's no harm in having them being set separately by different threads. 
2. I originally was looking at merge-with, but decided against it specifically due to the unnecessary complexity for no benefit. Simpler code is less volatile.

File: src/aging_session/memory.clj (hunk 7 above)
Comment (Elknar) [reply src/aging_session/memory.clj]: will do

File: src/aging_session/memory.clj (hunk 8 above)
Comment (Elknar) [reply src/aging_session/memory.clj]: can do... I was thinking of maybe plugging in some logging there actually though. similar to what we did with saml

File: src/aging_session/memory.clj (hunk 11 above)
Comment (Elknar) [reply src/aging_session/memory.clj]: it was key as in map key, since all the sessions are in a map... but yeah, can be confusing. 
can do session-id

File: src/aging_session/memory.clj (hunk 10 above)
Comment (Elknar) [reply src/aging_session/memory.clj]: when we issue a new session, which happens either when we issue a session for the first time; or when the presented cookie is invalid (corrupt/expired/tampered with) - decrypted-cookie is nil for all invalid cookies

File: test/aging_session/memory_test.clj (hunk 12 above)
Comment (Elknar) [reply test/aging_session/memory_test.clj]: (should probly be resolved in the above thread)
((this is copy-paste of an internal test where it has slightly more relevance - see other PR))

File: project.clj (hunk 2 above)
Comment (MatthewDarling) [reply project.clj:1]: That's kinda goofy to me. Have we never done breaking changes in a `0.X` -> `0.X+1` bump?

File: src/aging_session/memory.clj (hunk 5 above)
Comment (debora-duarte) [reply src/aging_session/memory.clj]: I'm here thinking if there will be a situation in which we may want to change one key and not the other 🤔 
If it makes sense to split this ```if``` check in two ```if``` checks... What do you think?

File: src/aging_session/memory.clj [hunk 13]
```diff
@@ -43,6 +64,67 @@
...
+        (swap! internals-atom update-in [:enc-key] #(if-not % (nonce/random-bytes 32) %))
+        (swap! internals-atom update-in [:mac-key] #(if-not % (nonce/random-bytes 128) %))
+        (get-crypto-keys)))))                               ; retrieve the keys again, as something else may have set them in the meantime
+
+(defn decrypt
+  [iv key data]
+  (try
+    (nippy/thaw (crypto/decrypt-cbc (crypto/block-cipher :aes :cbc) data key iv))
+    (catch Exception e
+      (log/error "Error decrypting cookie: " e)
```
Comment (debora-duarte) [inline src/aging_session/memory.clj]: Should we really be just logging it and just silently returning ```nil``` ?

File: src/aging_session/memory.clj (hunk 7 above)
Comment (debora-duarte) [reply src/aging_session/memory.clj]: One of the things we need to aim for in code that is shared is readability. 
I don't think the docstring is the answer here. The logic is good, but I think extracting some functions from there and maybe using the ```some->``` to go from one step to the other - verify, decrypt, check expiration - would make the code easier to read and understand and would make the docstring unnecessary.

Comment (debora-duarte) [reply src/aging_session/memory.clj]: Also, can you use different names for the variables? When you use `enc-cookie` for more than one purpose, it gives the feeling you're trying to mutate things - even if you're not mutating them - and it's kinda unnatural to do it like that in Clojure.
The approach of breaking stuff in more functions and using threading macros would also eliminate the need for more variable names

File: src/aging_session/memory.clj [hunk 14]
```diff
@@ -43,6 +64,67 @@
...
+  (let [{enc-key :enc-key
+         mac-key :mac-key} @internals-atom]
+    (if (and enc-key mac-key)                               ; if keys are present - use them
+      [enc-key mac-key]
+      (do                                                   ; otherwise generate new keys and attempt to set them
+        (swap! internals-atom update-in [:enc-key] #(if-not % (nonce/random-bytes 32) %))
+        (swap! internals-atom update-in [:mac-key] #(if-not % (nonce/random-bytes 128) %))
+        (get-crypto-keys)))))                               ; retrieve the keys again, as something else may have set them in the meantime
+
+(defn decrypt
```
Comment (debora-duarte) [inline src/aging_session/memory.clj]: Also: can you extract the logic for creating the cookie to another namespace? It seems it's better in terms of separation of concerns

**Review by debora-duarte — CHANGES_REQUESTED** [review]
Still need to take a look at the tests. Will do that ASAP

File: src/aging_session/memory.clj (hunk 5 above)
Comment (Elknar) [reply src/aging_session/memory.clj]: splitting the if wouldn't change behavior. we don't change keys ever here, only nil->key once for each. 

we don't have a session-store invalidation mechanic (only per-session delete), so no key change supported - just nuke the entire session-store and recreate it. 

there's also no particular reason to only change one of the keys. changing one invalidates all of the existing cookies, so may as well change the other.

File: src/aging_session/memory.clj (hunk 13 above)
Comment (Elknar) [reply src/aging_session/memory.clj]: what's the alternative? returning nil denies access - we treat wrong cookie as no cookie.

File: src/aging_session/memory.clj (hunk 14 above)
Comment (Elknar) [reply src/aging_session/memory.clj]: will do

File: src/aging_session/memory.clj (hunk 5 above)
Comment (debora-duarte) [reply src/aging_session/memory.clj]: Hum, ok. You're right

File: src/aging_session/memory.clj (hunk 7 above)
Comment (debora-duarte) [reply src/aging_session/memory.clj]: Regarding your question on Slack about the example I gave for `enc-cookie` : what I meant is not exactly more than one purpose. Yeah, it always represents the encrypted cookie, so it does not have more than one purpose.

What I meant is that binding a value to the same name again and again is seen as not idiomatic in Clojure and functional programming as a whole.
Clojure lets you do it - I didn't even know it could be done because more purely functional languages don't! - but this is considered bad practice. Clojure struggles to maintain mutability to a certain part of it and even has structures - like atoms - that are made specifically for cases where you want to mutate stuff. 
So doing:
```
(let [x 1
      x (+ 1 x)]
        (println x))
``` 
Is considered bad practice and is not even allowed in more purely functional languages. And Clojure let's you do it - God knows why! - but the idea is not to from a functional point of view.
What you should be doing is:

```
(let [x1 1
      x2 (+ 1 x1)]
        (println y))
``` 
Avoiding mutability in functional programming when it's permitted - like in Clojure - is also done because it helps you write a code that is easier to debug. In this second example of code, you could inspect the intermediate steps of your calculation more easily than in the first one example.

There's a similar discussion about it here:
https://softwareengineering.stackexchange.com/questions/306889/if-you-can-use-def-to-redefine-variables-how-is-that-considered-immutable

You can also look for "Imutability in functional programming" to understand more about it. But this is one of the main differences between the Functional Programming Paradigm vs Object Oriented or Imperative Paradigms.

File: src/aging_session/memory.clj (hunk 13 above)
Comment (debora-duarte) [reply src/aging_session/memory.clj]: What I'm thinking is that this is a "deal breaker" for the rest of the code, right? If this step fails, what's the point in doing anything else?

Catching and treating exceptions is something that makes more sense to be done when a step that is not mandatory fails or you can do something to recover from that - for instance, maybe get the value somewhere else? But this is neither the case here.

Will also be less obvious when debugging if things simply explode. 

Plus, you will be storing and accessing the database more with this implementation, because you'll be reading and storing things to it when it's not necessary. The `nil` does not short-circuit that. If you use something that bills you for access, you will be billed more because of that.

Comment (Elknar) [reply src/aging_session/memory.clj]: >If this step fails, what's the point in doing anything else?

we still have to process the response. an invalid cookie is just that - an invalid cookie. 

we shouldn't throw random 500s for invalid cookies - that's not user friendly and would only be confusing. 

processing an invalid cookie as no cookie makes more sense to me. also removes issues with users (if a non-malicious user somehow manages to have a corrupt cookie, throwing a 500 would not reset it for them. treating them as having no cookie and issuing a new one would)

File: test/aging_session/memory_test.clj [hunk 15]
```diff
@@ -2,54 +2,81 @@
...
+  (testing "encryption roundtrip"
```
Comment (debora-duarte) [inline test/aging_session/memory_test.clj]: cookie-test namespace? :please:

File: src/aging_session/memory.clj [hunk 16]
```diff
@@ -53,22 +54,27 @@
...
                     bump-session? (assoc :last-access (System/currentTimeMillis))
                     true (dissoc :bump-session?))]
       (swap! req-count inc)	  ; Increase the request count
       (if refresh-on-write    ; Write key and and update timestamp.
-        (swap! session-map assoc key (new-entry session))
-        (swap! session-map write-entry key session))
-      key))
+        (swap! session-map assoc return-id (new-entry session))
+        (swap! session-map write-entry return-id session))
+      (if session-id
```
Comment (debora-duarte) [inline src/aging_session/memory.clj:75]: It's not something that needs to be changed. But here you could have returned `(or cookie (cookie/bake-cookie return-id))`

File: test/aging_session/event_test.clj [hunk 17]
```diff
@@ -3,38 +3,41 @@
...
-    [aging-session.memory :refer :all]))
+    [aging-session.memory :refer :all]
+    [aging-session.cookie :refer :all]))
 
 (deftest session-expiry
   (testing "Test session expiry."
-    (let [as (aging-memory-store :events [(event/expires-after 1)])]
-      (write-session as "mykey" {:foo 1})
+    (let [as (aging-memory-store :events [(event/expires-after 1)])
+          cookie (write-session as "invalid-key" {:a 1})
```
Comment (debora-duarte) [inline test/aging_session/event_test.clj]: Why "invalid-key" as the id here?

File: test/aging_session/memory_test.clj [hunk 18]
```diff
@@ -2,54 +2,66 @@
...
 
 (deftest basic-read-empty
-  (testing "Test session reads."
-    (let [as (aging-memory-store)]
-      (is (= (read-session as "mykey") {})))))
+  (let [as (aging-memory-store)
+        cookie (write-session as "invalid-key" {:some :thing})]
+    (testing "Invalid session is empty."
+      (is (= (read-session as "invalid-cookie") {})))
+    (testing "Session with corrupt IV is empty"
```
Comment (debora-duarte) [inline test/aging_session/memory_test.clj]: I think these tests should be on the cookie-test namespace too

File: project.clj (hunk 1 above)
Comment (debora-duarte) [reply project.clj:13]: I believe that's the library we use for every crypto related stuff on other apps already. For what @Elknar told me, it's just a wrapper around Java libraries used to that

File: project.clj (hunk 2 above)
Comment (debora-duarte) [reply project.clj:1]: I don't have strong feelings about it. So I'm moving on with that ok @MatthewDarling ?

**Review by debora-duarte — APPROVED** [review]

**Comment by Elknar** [comment]
aaand off it goes
